modules/fedimint-ln-client/src/pay.rs

```diff
@@ -0,0 +1,545 @@
+        self.common.operation_id
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct LightningPayCreatedOutgoingLnContract {
+    pub funding_txid: TransactionId,
+    pub contract_id: ContractId,
+    pub gateway: LightningGateway,
+}
+
+impl LightningPayCreatedOutgoingLnContract {
+    fn transitions(
+        &self,
+        common: &LightningPayCommon,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<LightningPayStateMachine>> {
+        let txid = self.funding_txid;
+        let contract_id = self.contract_id;
+        let funded_common = common.clone();
+        let rejected_common = funded_common.clone();
+        let success_context = global_context.clone();
+        let reject_context = global_context.clone();
+        let gateway = self.gateway.clone();
+        vec![
+            // Success case: funding transaction is accepted
+            StateTransition::new(
+                Self::await_outgoing_contract_funded(funded_common.clone(), success_context, txid),
+                move |_dbtx, (), old_state| {
+                    Box::pin(Self::transition_outgoing_contract_funded(
+                        old_state,
+                        funded_common.clone(),
+                        contract_id,
+                        gateway.clone(),
+                    ))
+                },
+            ),
+            // Failure case: funding transaction is rejected
+            StateTransition::new(
+                Self::await_funding_transaction_rejected(
+                    rejected_common,
+                    reject_context,
+                    self.funding_txid,
+                ),
+                |_dbtx, (), old_state| Box::pin(Self::transition_canceled(old_state)),
+            ),
+        ]
+    }
+
+    async fn await_outgoing_contract_funded(
+        common: LightningPayCommon,
+        global_context: DynGlobalClientContext,
+        txid: TransactionId,
+    ) {
+        global_context
+            .await_tx_accepted(common.operation_id, txid)
+            .await;
+    }
+
+    async fn transition_outgoing_contract_funded(
+        old_state: LightningPayStateMachine,
+        common: LightningPayCommon,
+        contract_id: ContractId,
+        gateway: LightningGateway,
+    ) -> LightningPayStateMachine {
+        assert!(matches!(
+            old_state.state,
+            LightningPayStates::CreatedOutgoingLnContract(_)
+        ));
+
+        let payload = PayInvoicePayload::new(common.federation_id, contract_id);
+
+        LightningPayStateMachine {
+            common: old_state.common,
+            state: LightningPayStates::Funded(LightningPayFunded { payload, gateway }),
+        }
+    }
+
+    async fn await_funding_transaction_rejected(
+        common: LightningPayCommon,
+        global_context: DynGlobalClientContext,
+        txid: TransactionId,
+    ) {
+        global_context
+            .await_tx_rejected(common.operation_id, txid)
```

Comment:
I think we should refactor the await accepted/rejected API in the global context the same way I did for the client itself (1 fn returning a result):

https://github.com/fedimint/fedimint/blob/094f32c2ee2c675a04b1b53c49a6bd87fb38e51c/fedimint-client/src/lib.rs#L707-L722

That would allow merging the two state transitions, requiring fewer clones of notifiers etc.

Reply:
Implemented